crates/ark/src/lsp/diagnostics.rs

```diff
@@ -1094,4 +1218,144 @@ foo
+                x
+            ";
+
+            let document = Document::new(code, None);
+
+            let diagnostics = generate_diagnostics(document.clone(), DEFAULT_STATE.clone());
+            assert_eq!(diagnostics.len(), 1);
+
+            let diagnostic = diagnostics.get(0).unwrap();
+            insta::assert_snapshot!(diagnostic.message);
+        })
+    }
+
+    #[test]
+    fn test_dotty_assignment_recursive() {
+        r_test(|| {
+            let code = "
+                .[apple, .[banana]] <- list(1, list(2))
+                apple
+                banana
+                cherry
+            ";
+
+            let document = Document::new(code, None);
+
+            let diagnostics = generate_diagnostics(document.clone(), DEFAULT_STATE.clone());
+            assert_eq!(diagnostics.len(), 1);
+
+            let diagnostic = diagnostics.get(0).unwrap();
+            insta::assert_snapshot!(diagnostic.message);
+        })
+    }
+
+    #[test]
+    fn test_dotty_assignment_within_magrittr_pipe_braced_expr() {
+        r_test(|| {
+            let code = "
+                mtcars %>% list({ .[apple] <- 1; apple })
+                apple
+            ";
+
+            let document = Document::new(code, None);
+
+            let diagnostics = generate_diagnostics(document.clone(), DEFAULT_STATE.clone());
+            assert_eq!(diagnostics.len(), 1);
+
+            let diagnostic = diagnostics.get(0).unwrap();
+            assert_eq!(diagnostic.range.start.line, 2);
+            insta::assert_snapshot!(diagnostic.message);
+        })
+    }
+
+    #[test]
+    fn test_dotty_assignment_within_native_pipe_braced_expr() {
+        // TODO: `apple` should be defined in the global env and there should not be a diagnostic here
+        r_test(|| {
+            let code = "
+                mtcars |> list({ .[apple] <- 1; apple })
+                apple
+            ";
+
+            let document = Document::new(code, None);
+
+            let diagnostics = generate_diagnostics(document.clone(), DEFAULT_STATE.clone());
+            assert_eq!(diagnostics.len(), 1);
+
+            let diagnostic = diagnostics.get(0).unwrap();
+            assert_eq!(diagnostic.range.start.line, 2);
+            insta::assert_snapshot!(diagnostic.message);
+        })
+    }
+
+    #[test]
+    fn test_assignment_within_function_arguments() {
+        // TODO: `x` should be defined in the global env and there should not be a diagnostic here
+        r_test(|| {
+            let code = "
+                list({ x <- 1 })
+                x
+            ";
+
+            let document = Document::new(code, None);
+
+            let diagnostics = generate_diagnostics(document.clone(), DEFAULT_STATE.clone());
+            assert_eq!(diagnostics.len(), 1);
+
+            let diagnostic = diagnostics.get(0).unwrap();
+            assert_eq!(diagnostic.range.start.line, 2);
+            insta::assert_snapshot!(diagnostic.message);
+        })
```

Comment:
We have kind of a general issue here, so I've recorded this "supposed to be failing" test. Definitely a tough one though, i.e. NSE really throws this off.